[CHANGED] Randomize Leafnode remote URLs and add option to disable

variadico opened the pull request:
Currently, leafnode remote URLs are kept in order when parsed. This change randomizes the order of URLs by default and adds an option to disable this behavior, returning you to the way it was before.

```
leafnodes {
  remotes = [
    {
      dont_randomize: true
      url: [
        "nats-leaf://host1:7422"
        "nats-leaf://host2:7422"
        "nats-leaf://host3:7422"
        "nats-leaf://host4:7422"
      ]
      account: fizzbuzz
      credentials: "./fizzbuzz.creds"
    }
  ]
}
```

Resolves #2153

/cc @nats-io/core

kozlovic commented on server/opts.go:
```diff
@@ -1842,6 +1846,11 @@ func parseRemoteLeafNodes(v interface{}, errors *[]error, warnings *[]error) ([]
 			}
 		}
+		if remote.ShuffleURLs {
+			rand.Shuffle(len(remote.URLs), func(i, j int) {
```
I would not do the shuffling in parsing of options (which would make your test to have to be updated). The reason is that users embedding the server would not benefit from that if they configure the server directly with options.
I recommend doing checks/validations in code that is done in NewServer() which then means it applies to both config and direct options. For leafnode, this would be `validateLeafNode()`.

variadico replied:
Aaah, ok. Great thanks, will take another look.

kozlovic commented on server/opts_test.go:
```diff
@@ -2383,6 +2384,127 @@ func TestParsingLeafNodeRemotes(t *testing.T) {
 		}
 	})
+
+	t.Run("shuffle remote urls no effect", func(t *testing.T) {
```
Unfortunately, you would have to change those tests if you move the shuffling. You would even have to add test with direct options (so 4 tests total): 2 with ProcessConfigFile() followed by NewServer() and then check remotes' urls array, and 2 with creation Options and setting options directly and then calling NewServer() and checking remotes.
Let me know if you need more details on that.

matthiashanel commented:
@variadico @kozlovic , wondering if we should enable this by default and provide an option to turn it off. In part because this is the right thing to do, Why make it harder by having to enable the better behavior.

ripienaar commented:
Agree. Should be default.

kozlovic commented:
Yes, sorry I missed that too. Since I saw the discussion on Slack and issue reported by Don R., I assumed that this was the direction we would take. @variadico The flag would have to be a negative, like "NoSuffle" or something like that.

variadico commented:
All right. Got some updates.

* Moved `rand.Shuffle` to `validateLeafNode`
* Made `rand.Shuffle` the default behavior
* Changed the name "shuffle" to "random" to match the term the NATS client uses (any thoughts on this?)
* Updated the parsing code
* Added new tests

kozlovic reviewed (approved):
LGTM!
